lib/debug/thread_client.rb

```diff
@@ -359,21 +359,27 @@ def frame_eval src, re_raise: false
       end
     end
 
-    def frame_str(i)
+    def frame_str(i, frames: @target_frames)
       cur_str = (@current_frame_index == i ? '=>' : '  ')
       prefix = "#{cur_str}##{i}"
-      frame = @target_frames[i]
+      frame = frames[i]
       frame_string = @frame_formatter.call(frame)
       "#{prefix}\t#{frame_string}"
     end
 
-    def show_frames max = (@target_frames || []).size
+    def show_frames(max = (@target_frames || []).size, location_pattern: nil)
       if max > 0 && @target_frames
-        size = @target_frames.size
-        max += 1 if size == max + 1
+        frames =
+          if location_pattern
+            @target_frames.select { |f| f.location_str.match?(location_pattern) }
+          else
+            @target_frames
+          end
+
+        size = frames.size
         max.times{|i|
           break if i >= size
-          puts frame_str(i)
+          puts frame_str(i, frames: frames)
```

Comment:
Before filter

```
#0: foo
#1: bar
#2: boo
#3: baz
```

with filter /a/

```
#0: bar
#1: baz
```

But it is another option:

```
#1: bar
#3: baz
````

(remain original index).

Pros:
* We can recognize how deep frames.
Cons:
* Need to input huge number like 123 even if only 4 frames are shown.

Reply:
I prefer remaining original index. But not sure how users feel...

Reply:
I think it provides better information to users 👍 
But this requires assigning the frame index to `FrameInfo`, which is a bit outside the scope of this PR. Can I open another one for this?

Reply:
You don't need to store the index into FrameInfo, but you need to collect with index like,

```ruby
fs = []
frames.each_with_index([]){|f, i|
  r << [i, f] if f.match(pat)
}```

Reply:
ah, and we can choose this frame with `frame` command with original index.

Reply:
I see 🤔 I'll open another PR for this after this one is merged.